app/assets/javascripts/govuk_publishing_components/components/input.js

```diff
@@ -0,0 +1,54 @@
+window.GOVUK = window.GOVUK || {}
+window.GOVUK.Modules = window.GOVUK.Modules || {};
+
+(function (Modules) {
+  function Input () { }
+
+  Input.prototype.start = function ($module) {
+    this.$module = $module[0]
+    this.input = this.$module.querySelector('.gem-c-input')
+    this.$module.togglePassword = this.togglePassword.bind(this)
+    this.$module.revertToPasswordOnFormSubmit = this.revertToPasswordOnFormSubmit.bind(this)
+
+    this.showPasswordText = this.$module.getAttribute('data-show')
+    this.hidePasswordText = this.$module.getAttribute('data-hide')
+    this.shownPassword = this.$module.getAttribute('data-announce-show')
+    this.hiddenPassword = this.$module.getAttribute('data-announce-hide')
+
+    this.showHide = document.createElement('button')
+    this.showHide.className = 'gem-c-input__password-toggle gem-c-button govuk-button gem-c-button--secondary-quiet'
+    this.showHide.setAttribute('aria-controls', this.input.getAttribute('id'))
+    this.showHide.innerHTML = this.showPasswordText
+    this.$module.insertBefore(this.showHide, this.input.nextSibling)
+
+    this.statusText = document.createElement('span')
+    this.statusText.classList.add('govuk-visually-hidden')
+    this.statusText.innerHTML = this.hiddenPassword
+    this.statusText.setAttribute('aria-live', 'polite')
+    this.$module.insertBefore(this.statusText, this.showHide.nextSibling)
+
+    this.showHide.addEventListener('click', this.$module.togglePassword)
+
+    this.parentForm = this.input.form
+    var disableFormSubmitCheck = this.$module.getAttribute('data-disable-form-submit-check')
+
+    if (this.parentForm && !disableFormSubmitCheck) {
+      this.parentForm.addEventListener('submit', this.$module.revertToPasswordOnFormSubmit)
```

Comment:
It looks like this could accidentally add quite a few event handlers to a form. If I understand right this will add a submit event handler for every input component whether or not they are for a password.

Do you know if this definitely solves a problem or is a precaution?

A way around it could be a static function added to the eventListener so it doesn't add this multiple times, or to only apply this when it's a password input.

Reply:
The JS should only initialise for an input with the password reveal turned on, not for every instance of the input component. The data-module attribute should only be present when the component is set to reveal password mode.

It's unlikely that anyone would use this more than twice in a form (you might have password and confirm password) so I'm not too worried about excessive event handlers.

Reply:
Oh is that because only this usage of the component has the data-module attribute of input? I didn't catch that. It seems like that might be a little deceptive if the whole JS for the component is only for a particular modifier. Perhaps that would be clearer if that JS initialised for all instances but exited early when not for password.